Refactor youtube upload script

lightandluck opened the pull request:
This addresses comments from https://github.com/edgi-govdata-archiving/edgi-scripts/pull/55.

We organized the files so that it would be easier to differentiate between library modules and script modules. Also, updated the README.

One issue - We added constants for Youtube video categories and playlists. The ID for our unlisted playlist was added as an environment variable similar to Zoom keys. I feel like I used a bad pattern in how I set the constant using env variables, so would appreciate a second look there.

lightandluck committed:
Add video_category_ID constants

lightandluck committed:
Delete unused files

lightandluck committed:
Add description to archive.sh

lightandluck committed:
Separate library modules from runnable modules

lightandluck committed:
Refactor youtube upload according to comments from https://github.com/edgi-govdata-archiving/edgi-scripts/pull/55

lightandluck committed:
Add playlist IDs as constants so we don't have to look them up using YouTube API everytime

lightandluck committed:
Replace print statements with debug statements in youtube module

lightandluck committed:
Uncomment filter video by length

lightandluck committed:
Update README

lightandluck committed:
Update README

lightandluck commented on scripts/lib/basic_youtube_upload.py:
```diff
@@ -32,18 +35,6 @@
-#   https://developers.google.com/youtube/v3/guides/authentication
-# For more information about the client_secrets.json file format, see:
-#   https://developers.google.com/api-client-library/python/guide/aaa_client_secrets
-CLIENT_SECRETS_FILE = 'client_secret.json'
```
This file is actually not needed here, but is used instead when running `scripts/auth.py`

lightandluck commented on scripts/lib/constants.py:
```diff
@@ -0,0 +1,50 @@
+    'Data Together': 'PLtsP3g9LafVul1gCctMYGm9sz5FUWr5bu',
+    'Website Monitoring': 'PLtsP3g9LafVtj6IOMk05aOh-DdpuqWhue',
+    'Community Calls': 'PLtsP3g9LafVsaa18lQaPXzxJU7wIcPB1O',
+    'Uploads from Zoom': os.environ['DEFAULT_YOUTUBE_PLAYLIST']
```
This feels really brittle to me. Not sure what a better pattern would be...

Mr0grog replied:
Yeah, I would probably not put this here. My tendency would be to either:

- Just let `add_video_to_playlist()` do its thing and use the API to look up the playlist (it’s pretty cheap) rather than needing the ID.

- Look up the env var for this in the main script, and only add to the default if present. e.g. in `upload_zoom_recordings.py`, something like:

    ```py
    default_playlist = os.environ.get('DEFAULT_YOUTUBE_PLAYLIST')

    # ...and down where you upload the video...
    video_id = upload_video(blah, blah, blah)
    if default_playlist:
        print('  Adding to main playlist: Uploads from Zoom')
        add_video_to_playlist(youtube, video_id, title=default_playlist,
                              privacy='unlisted')
        # Or if the env var is an ID instead of a name:
        print('  Adding to main playlist: Uploads from Zoom')
        add_video_to_existing_playlist(youtube, video_id, default_playlist)
    ```

Mr0grog commented on scripts/lib/basic_youtube_upload.py:
```diff
@@ -52,13 +43,43 @@
 # Create client from stored authorization credentials.
-def get_youtube_client():
-    credentials = google.oauth2.credentials.Credentials.from_authorized_user_file(YOUTUBE_CREDENTIALS_FILE)
+def get_youtube_client(youtube_credentials):
```
I might consider renaming this parameter to clarify that it should be a file path, e.g. `credentials_path`.

Mr0grog commented on scripts/lib/basic_youtube_upload.py:
```diff
@@ -52,13 +43,43 @@
-def initialize_upload(youtube, file, title='Test Title', description=None,
+def upload_video(youtube, file, title='Test Title', description=None,
                       category=None, tags=None, privacy_status='private',
                       recording_date=None, license=None):
```
The indentation here should be fixed to line up with the `(`. [Pyflakes](https://pypi.org/project/pyflakes/) (more limited) or [Flake8](https://pypi.org/project/flake8/) (more style checks) are good linters that can also help here!

Mr0grog commented on scripts/lib/basic_youtube_upload.py:
```diff
@@ -130,20 +151,84 @@ def resumable_upload(request):
+def get_known_playlist(title):
+    return PLAYLIST_IDS[title]
+
+def get_playlist(youtube, title):
```
*Really minor* nit: I think `find_playlist_id` might be a better name here: `find_` since you are really looking for a playlist based on some criteria (rather than getting a known playlist by its ID) and `_id` since you are getting the ID, not the playlist data. This isn’t a huge deal though.

I think it also might be worthwhile to split this method in two since you’re refactoring:

1. `list_playlists`/`get_playlists` (or whatever name you like) that just produces an iterable sequence of all a user’s playlists.

2. This method would then just call that filters that list:

    ```py
    def find_playlist_id(youtube, title):
        for playlist in list_playlists(youtube):
            if title == playlist.get('snippet', {}).get('title'):
                return playlist.get('id')
    ```

Mr0grog commented on scripts/lib/basic_youtube_upload.py:
```diff
@@ -130,20 +151,84 @@ def resumable_upload(request):
+            debug('Sleeping %f seconds and then retrying...' % sleep_seconds)
             time.sleep(sleep_seconds)
 
+def get_known_playlist(title):
```
Similar note as below: I think it might be good to tack `_id` onto the end of the name here since you’re getting the ID, not the playlist data.

Mr0grog commented on scripts/lib/basic_youtube_upload.py:
```diff
@@ -130,20 +151,84 @@ def resumable_upload(request):
             time.sleep(sleep_seconds)
 
+def get_known_playlist(title):
+    return PLAYLIST_IDS[title]
```
This will raise a `KeyError` if `title` doesn’t exist in `PLAYLIST_IDS`. Based on how you’re using this function, you probably want to use [`dict.get()`](https://docs.python.org/3.6/library/stdtypes.html#dict.get):

```suggestion
    return PLAYLIST_IDS.get(title)
```

Mr0grog commented on scripts/lib/basic_youtube_upload.py:
```diff
@@ -130,20 +151,84 @@ def resumable_upload(request):
+        results = request.execute()
+        for item in results["items"]:
+            t = item.get("snippet", {}).get("title")
+            existing_playlist_title = (t.encode(current_encoding) if hasattr(t, 'decode') else t)
```
I know this is just copied from the old `playlists` file (which came from edgi-govdata-archiving/youtube-upload), but I’m 100% sure this logic is wrong.

It probably came from Python 2, where it might’ve been OK. Strings in Py2 had both `encode` and `decode` methods because there was a complicated/ambiguous relationship between the `str` type, bytestrings, and unicode strings. In Py3 things are more clearly separated, and the `str` type is always a unicode string (which *only* has `encode()`) and `bytes` is a separate type (which *only* has `decode()`). Since we’re testing for `decode()` but then calling `encode()`, this would probably always break in Py3 (which we are always using).

That *also* means that `t` is always a normal unicode `str` type now (since we’d be breaking if `hasattr(t, 'decode')` were true). I think we can probably safely remove this encoding logic. (It was added [4 years ago](https://github.com/edgi-govdata-archiving/youtube-upload/commit/a22db3af5fbacc35cdb941b96fbbb76de9b3e189), after all!)

Mr0grog commented on scripts/lib/basic_youtube_upload.py:
```diff
@@ -130,20 +151,84 @@ def resumable_upload(request):
+
+def create_playlist(youtube, title, privacy):
+    """Create a playlist by title and return its ID"""
+    debug("Creating playlist: {0}".format(title))
```
Better to use an f-string for this these days:

```suggestion
    debug(f"Creating playlist: {title}")
```

(You could probably do this all the places `%` interpolation is used, too, but ¯\\\_(ツ)\_/¯)

Mr0grog commented on scripts/lib/basic_youtube_upload.py:
```diff
@@ -130,20 +151,84 @@ def resumable_upload(request):
+
+def add_video_to_existing_playlist(youtube, playlist_id, video_id):
+    """Add video to playlist (by identifier) and return the playlist ID."""
+    debug("Adding video to playlist: {0}".format(playlist_id))
```
Same here w/r/t f-strings.

Mr0grog commented on scripts/lib/basic_youtube_upload.py:
```diff
@@ -130,20 +151,84 @@ def resumable_upload(request):
+    """Add video to playlist (by title) and return the full response."""
+    playlist_id = get_known_playlist(title) or \
+        get_playlist(youtube, title) or \
+        create_playlist(youtube, title, privacy)
```
FYI, if you wrap this expression in parens, you can drop the trailing `\`:

```suggestion
    playlist_id = (get_known_playlist(title) or
                   get_playlist(youtube, title) or
                   create_playlist(youtube, title, privacy))
```

Both ways are totally fine, I just happen to think the parentheses make for *slightly* nicer style. Totally personal preference.

Mr0grog commented on scripts/lib/constants.py:
```diff
@@ -0,0 +1,50 @@
+        'basic': 1,
+        'pro': 2,
+        'corp': 3,
+        }
```
Minor formatting nit… either line things up with the brackets:

```suggestion
USER_TYPES = {'basic': 1,
              'pro': 2,
              'corp': 3 }
```

Or use standard 4-space indentation:

```suggestion
USER_TYPES = {
    'basic': 1,
    'pro': 2,
    'corp': 3,
}
```

Mr0grog commented on scripts/upload_zoom_recordings.py:
```diff
@@ -87,9 +84,7 @@ def download_file(url, download_path, query=None):
-
-youtube = get_youtube_client()
+youtube = get_youtube_client(YOUTUBE_CREDENTIALS_FILE)
 with tempfile.TemporaryDirectory() as tmpdirname:
```
While you’re here refactoring, it can be helpful to wrap this code in a function and call it only if being loaded as a script rather than imported as a library:

```py
def main():
    # All the main program logic here

if __name__ == '__main__':
    main()
```

That way it’s easier to write tests later, because you can `import` this module in your tests without causing the main program logic to immediately execute.

Mr0grog reviewed (commented):
This is a great improvement! I left a lot of little notes about small nits and further improvements inline, but it’s already a big step forward.

Finally, I think it might be good to rename `basic_youtube_upload.py` to just `youtube.py` or something, since `basic` isn’t really differentiating it from something else, and the module now covers a lot more than just uploading. :)

lightandluck committed:
Rename variables

lightandluck committed:
Rename functions and place upload in main()

lightandluck committed:
Change format strings

lightandluck committed:
Update README

lightandluck committed:
Remove import

lightandluck committed:
Rename parameter

lightandluck committed:
Uncomment video length filter

lightandluck commented:
Hi @Mr0grog, so went in and addressed the suggestions. Good stuff, thanks for pointing out so many things. I went ahead and just removed all the new logic with finding playlist ids. It was just adding more unneeded complexity. 

I installed flake8 as a linter and it came back with a ton of messages! A lot of them were about spacing and lines being too long. A couple unused variables and libraries too. I think I'll address those in another PR to keep the noise down in this one. Thanks for the review 🙌

lightandluck committed:
Remove unused variable in README
